### PR TITLE
Store salted and hashed passwords in the database

### PR DESCRIPTION
**Pull request by SamWhited**
I do not use MySQL, so I have only tested this with PostgreSQL. This PR does not include any mechanism for migrating existing installs to the new schema or updating existing passwords.

### CODE REVIEW

**Comment by SamWhited** [comment]
Rebased, hopefully that's the last one. Let me know if you find any issues; thanks!

**Comment by SamWhited** [comment]
Gentle ping since it's been a week; of course, if you're just busy and can't get to this no rush, I just wanted to make sure it doesn't drop off the bottom of your list. Thanks!

**Comment by SamWhited** [comment]
Ping; is this waiting on anything? Thanks!

File: model/user.go [hunk 1]
```diff
@@ -15,10 +15,13 @@ import (
...
 type User struct {
-	Username       string
-	Password       string
-	LastPresence   *xmpp.Presence
-	LastPresenceAt time.Time
+	Username            string
+	PasswordScramSHA1   []byte
+	PasswordScramSHA256 []byte
+	Salt                []byte
+	IterationCount      int
```
Comment (mmalesev) [inline model/user.go:22]: Perhaps it would be cleaner to have this is a separate Password struct

Comment (SamWhited) [reply model/user.go:22]: I feel like the extra nesting would just make the code uglier, but I have no strong preference if the maintainer likes it one way or another.

**Comment by ortuman** [comment]
🙏  Sorry for the late response. Just wanted to let you know that currently I'm working on a redesign of jackal core which I expect to have ready in the coming months. As part of this rewrite I've introduced encrypted password (peppered and salted), so If you don't mind, I'd wait until new changes are available on master branch and then see if anything can be improved. Thanks for all fresh ideas you came up with. 🙌

**Pull request closed without merging**